zkevm-circuits/src/keccak_circuit/keccak_bit.rs

```diff
@@ -0,0 +1,870 @@
+                .collect::<Vec<_>>()[0..4]
+                .concat();
+
+            let rlc = compose_rlc(&hash_bytes, r);
+
+            cb.condition(meta.query_advice(q_end, Rotation::cur()), |cb| {
+                cb.require_equal(
+                    "hash rlc check",
+                    rlc,
+                    meta.query_advice(hash_rlc, Rotation::cur()),
+                );
+            });
+
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        println!("degree: {}", meta.degree());
+
+        KeccakBitConfig {
+            q_enable,
+            q_first,
+            q_round,
+            q_absorb,
+            q_end,
+            hash_rlc,
+            s_bits,
+            c_bits,
+            a_bits,
+            iota_bits,
+            c_table,
+            _marker: PhantomData,
+        }
+    }
+
+    fn assign(
+        &self,
+        mut layouter: impl Layouter<F>,
+        _size: usize,
+        witness: &[KeccakRow<F>],
+    ) -> Result<(), Error> {
+        layouter.assign_region(
+            || "assign keccak rounds",
+            |mut region| {
+                for (offset, keccak_row) in witness.iter().enumerate() {
+                    self.set_row(
+                        &mut region,
+                        offset,
+                        keccak_row.q_end,
+                        keccak_row.hash_rlc,
+                        keccak_row.s_bits,
+                        keccak_row.c_bits,
+                        keccak_row.a_bits,
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn set_row(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        q_end: u64,
+        hash_rlc: F,
+        s_bits: [u8; KECCAK_WIDTH],
+        c_bits: [u8; C_WIDTH],
+        a_bits: [u8; ABSORB_WIDTH_PER_ROW],
+    ) -> Result<(), Error> {
+        let round = (offset + 24) % 25;
+
+        self.q_enable.enable(region, offset)?;
+
+        // q_first
+        if offset == 0 {
+            self.q_first.enable(region, offset)?;
+        }
+
+        // q_round
+        region.assign_fixed(
+            || format!("assign q_round {}", offset),
+            self.q_round,
+            offset,
+            || Ok(F::from(round != 24)),
+        )?;
+        // q_absorb
+        region.assign_fixed(
+            || format!("assign q_absorb {}", offset),
+            self.q_absorb,
+            offset,
+            || Ok(F::from(round == 24)),
+        )?;
+        // q_end
+        region.assign_advice(
+            || format!("assign q_end {}", offset),
+            self.q_end,
+            offset,
+            || Ok(F::from(q_end)),
+        )?;
+        // hash_rlc
+        region.assign_advice(
+            || format!("assign hash_rlc {}", offset),
+            self.hash_rlc,
+            offset,
+            || Ok(hash_rlc),
+        )?;
```

Comment:
With Han's query_challenge, this won't even be a parameter that is passed. But we can keep it for now.

Reply:
> We need to take into account that https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/659 will imply that the RLC is not longer a witness we provide to the circuit and rather just a value we "compute and can pass over to other circuits" via lookups.

I guess I'm not really sure how to best write these things until the challenge API is available. For things like the RLC encoded input bytes that can stretch a lot of rows, how should that be done so it is easily convertible to the new API? Even the calculations of the RLC encoded value of the keccak hash itself (which is the value exposed above) can be quite complex, so not sure if just writing the code now like there is some const randomness value makes sense.

Reply:
The `query_challenge` API will return `Value<F>` for usage, we can simply mock it with a `F::random()` for now (or some static random value). So I think currently the way this PR implemented should be easy to adapt.